In [4]:
import pandas as pd
import numpy as np
import os
import glob
from collections import Counter
from __future__ import division

In [6]:
df_original = pd.read_table("C:\Users\Apourva\Downloads\two_month\cmds_sequence_2016-07-01.csv", delimiter="$$", header = None, engine='python')

IOError: [Errno 22] invalid mode ('r') or filename: 'C:\\Users\\Apourva\\Downloads\two_month\\cmds_sequence_2016-07-01.csv'

In [6]:
df_original.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 54857 entries, 0 to 54856
Data columns (total 1 columns):
0    54857 non-null object
dtypes: object(1)
memory usage: 428.6+ KB


In [62]:
df = df_original[0].apply(lambda x: pd.Series(x.split('$$')))

In [7]:
path = r'C:\Users\Apourva\Downloads\two_month'
allFiles = glob.glob(os.path.join(path, "*.csv"))
c = 0

for f in allFiles:
    print f
    d1 = pd.read_csv(f, delimiter="$$", header = None, engine='python')
    c = c + d1.shape[0]
    print c

C:\Users\Apourva\Downloads\two_month\cmds_sequence_2016-07-01.csv
54857
C:\Users\Apourva\Downloads\two_month\cmds_sequence_2016-07-02.csv
111769
C:\Users\Apourva\Downloads\two_month\cmds_sequence_2016-07-03.csv
166649
C:\Users\Apourva\Downloads\two_month\cmds_sequence_2016-07-04.csv
223895
C:\Users\Apourva\Downloads\two_month\cmds_sequence_2016-07-05.csv
281263
C:\Users\Apourva\Downloads\two_month\cmds_sequence_2016-07-06.csv
327711
C:\Users\Apourva\Downloads\two_month\cmds_sequence_2016-07-07.csv
383981
C:\Users\Apourva\Downloads\two_month\cmds_sequence_2016-07-08.csv
434513
C:\Users\Apourva\Downloads\two_month\cmds_sequence_2016-07-09.csv
478631
C:\Users\Apourva\Downloads\two_month\cmds_sequence_2016-07-10.csv
522445
C:\Users\Apourva\Downloads\two_month\cmds_sequence_2016-07-11.csv
565457
C:\Users\Apourva\Downloads\two_month\cmds_sequence_2016-07-12.csv
613236
C:\Users\Apourva\Downloads\two_month\cmds_sequence_2016-07-13.csv
667894
C:\Users\Apourva\Downloads\two_month\cmds_sequence_2

In [7]:
df_ip = pd.read_table("C:\Users\Apourva\Downloads\ips.txt", header = None)

In [8]:
df_ip.shape

(3414011, 1)

In [9]:
ip_list = df_ip[0].tolist()
ip_frequency_count = Counter(ip_list)

In [10]:
len(ip_list)

3414011

In [11]:
# Top 10 most frequently occuring IP addresses
ip_frequency_count.most_common(10)

[('115.177.11.215', 294690),
 ('115.176.182.196', 259241),
 ('192.3.106.42', 66572),
 ('191.96.249.189', 51251),
 ('104.192.0.20', 18707),
 ('198.204.234.26', 6817),
 ('198.204.234.27', 5284),
 ('198.204.234.28', 5239),
 ('185.93.185.10', 5157),
 ('198.204.234.30', 5049)]

In [14]:
def Frequent (k):
    c = {}
    T = []
    for ip in ip_list:
        if ip in T:
            c[ip] = c[ip] + 1
        else:
            if len(T) < k:
                T.append(ip)
                c[ip] = 1
            else:
                for j in T:
                    c[j] = c[j] - 1
                    if c[j] == 0:
                        T.remove(j)
    return T, c

In [15]:
T_10, c_10 = Frequent ( 10 )
for i in T_10:
    print i, " : " , c_10[i]

115.176.182.196  :  33891
123.23.14.2  :  1
222.105.38.62  :  1
187.199.79.13  :  2
187.23.203.254  :  1
87.96.167.236  :  1


According to the FREQUENT algorithm, all IP addresses which have frequency greater than n/k must be reported.
Here, n = 3414011 and k = 10 and n/k = 341401

The most commonly occuring IP address (115.177.11.215) only has frequency of 294690 which is less that 341401. Hence, we expect that the FREQUENT algorithm does not return any IP addresses.

However, we see that the algorithm returns 6 values.


In [16]:
T_100, c_100 = Frequent ( 100 )
print "length of T_100 : ", len(T_100)
for i in T_100:
    print i, " : ", c_100[i]


length of T_100 :  100
115.177.11.215  :  241575
192.3.106.42  :  23965
115.176.182.196  :  227238
191.96.249.189  :  26915
179.32.67.173  :  1
88.253.72.22  :  1
2.27.193.109  :  2
190.255.171.160  :  1
221.224.112.227  :  1
218.144.19.189  :  1
79.70.235.37  :  1
218.5.244.39  :  1
110.92.230.34  :  2
94.228.241.131  :  1
59.152.138.124  :  1
211.36.166.120  :  1
125.136.64.201  :  1
167.0.5.5  :  1
87.96.167.236  :  2
14.176.162.21  :  1
122.117.89.151  :  1
88.247.151.215  :  1
85.228.152.223  :  1
80.13.179.242  :  1
61.238.48.112  :  1
122.89.213.99  :  1
211.216.87.109  :  1
201.228.239.228  :  1
84.52.87.130  :  1
200.82.151.87  :  1
191.109.252.183  :  1
59.124.21.67  :  1
114.32.140.177  :  1
117.213.22.135  :  1
59.8.221.31  :  1
166.239.150.247  :  1
85.175.227.22  :  1
46.172.72.178  :  1
121.134.10.195  :  1
69.90.132.210  :  1
125.227.17.109  :  1
190.66.183.47  :  1
182.186.86.150  :  1
212.253.83.5  :  1
70.62.189.142  :  1
14.137.88.2  :  1
200.38.234.163  :  1
117.1.

In [133]:
dict_100 = {}
for i in ip_frequency_count:
    if ip_frequency_count[i] > 34140:
        dict_100[i] = ip_frequency_count[i]
print dict_100

{'192.3.106.42': 66572, '115.177.11.215': 294690, '191.96.249.189': 51251, '115.176.182.196': 259241}


According to the FREQUENT algorithm, all IP addresses which have frequency greater than n/k must be reported. Here, n = 3414011 and k = 100 and n/k = 34140
There are 6 IP addersses that must be identified based on the n/k bound

{'192.3.106.42': 66572, '115.177.11.215': 294690, '191.96.249.189': 51251, '115.176.182.196': 259241}

We see that the FREQUENT algorithm does indeed identify these IP addresses but also returns 96 other IP addresses whose frequencies are lower than 34140.

In [17]:
T_1000, c_1000 = Frequent ( 1000 )
print "length of T_1000 : ", len(T_1000)
for i in T_1000:
    print i, " : ", c_1000[i]

length of T_1000 :  761
115.177.11.215  :  289897
198.204.234.26  :  2243
104.192.0.20  :  14036
104.192.0.22  :  98
198.204.234.27  :  1028
198.204.234.28  :  986
198.204.234.29  :  636
198.204.234.30  :  797
192.3.106.42  :  62369
115.176.182.196  :  256150
191.96.249.189  :  48147
185.93.185.10  :  3911
2.27.193.109  :  382
191.80.232.207  :  9
62.23.168.212  :  317
124.105.61.164  :  2
191.109.131.25  :  2
124.106.155.77  :  15
91.189.252.130  :  6
190.255.171.160  :  49
186.119.226.23  :  36
188.226.188.75  :  36
86.105.50.213  :  51
69.90.132.210  :  23
190.67.234.147  :  45
211.111.237.37  :  44
161.18.80.235  :  46
179.32.67.173  :  46
201.228.239.228  :  42
186.115.252.0  :  39
186.116.109.40  :  15
110.92.230.34  :  34
190.255.84.59  :  44
186.170.48.227  :  45
122.52.229.171  :  48
167.0.12.94  :  32
111.179.36.22  :  39
201.76.71.177  :  39
191.109.252.183  :  43
119.92.247.209  :  26
187.199.79.13  :  40
186.115.109.168  :  49
190.66.183.47  :  46
167.0.5.5  :  41
179.33.6

In [18]:
dict_1000 = {}
not_identified_by_algorithm = []
for i in ip_frequency_count:
    if ip_frequency_count[i] > 3414:
        dict_1000[i] = ip_frequency_count[i]
        if i not in T_1000:
            not_identified_by_algorithm.append(i)
print dict_1000
print "length : " , len(dict_1000)
print "IP addresses not reported by FREQUENT alforithm : ", not_identified_by_algorithm

{'198.204.234.29': 4889, '198.204.234.28': 5239, '185.93.185.10': 5157, '104.192.0.18': 3701, '198.204.234.30': 5049, '198.204.234.27': 5284, '198.204.234.26': 6817, '192.3.106.42': 66572, '104.192.0.22': 4353, '115.29.251.112': 3812, '104.192.0.20': 18707, '115.176.182.196': 259241, '115.177.11.215': 294690, '191.96.249.189': 51251}
length :  14
IP addresses not reported by FREQUENT alforithm :  ['104.192.0.18', '115.29.251.112']


According to the FREQUENT algorithm, all IP addresses which have frequency greater than n/k must be reported. Here, n = 3414011 and k = 1000 and n/k = 3414

The 14 IP addresses which must be identified their respective counts are shown below 

{'198.204.234.29': 4889, '198.204.234.28': 5239, '185.93.185.10': 5157, '104.192.0.18': 3701, '198.204.234.30': 5049, '198.204.234.27': 5284, '198.204.234.26': 6817, '192.3.106.42': 66572, '104.192.0.22': 4353, '115.29.251.112': 3812, '104.192.0.20': 18707, '115.176.182.196': 259241, '115.177.11.215': 294690, '191.96.249.189': 51251}

We see that the FREQUENT algorithm identifies only 12 out of the 14 IPs. It also returns 745 IP addresses whose frequencies are lower than 3414.